File: MiniDOMOffice/Office/DefinizionePercorsi/LuoghiXPercorso/LuoghiDefinitiPerPercorso.cs
using System;
using System.Collections;
using System.Collections.Generic;
using DMD;
using DMD.XML;
using DMD.Databases;
using DMD.Databases.Collections;
using minidom;
using minidom.repositories;
using static minidom.Sistema;
using static minidom.Anagrafica;
using static minidom.Office;

namespace minidom
{
    public partial class Office
    {

        /// <summary>
        /// Collezione di luoghi da visitare per un percorso
        /// </summary>
        [Serializable]
        public class LuoghiDefinitiPerPercorso
            : CCollection<LuogoDaVisitare>
        {

            [NonSerialized] private PercorsoDefinito m_Percorso;

            /// <summary>
            /// Costruttore
            /// </summary>
            public LuoghiDefinitiPerPercorso()
            {
                m_Percorso = null;
            }

            /// <summary>
            /// Costruttore
            /// </summary>
            /// <param name="percorso"></param>
            public LuoghiDefinitiPerPercorso(PercorsoDefinito percorso) 
                : this()
            {
                Load(percorso);
            }

            /// <summary>
            /// Restituisce un riferimento al percorso
            /// </summary>
            public PercorsoDefinito Percorso
            {
                get
                {
                    return m_Percorso;
                }
            }

            /// <summary>
            /// OnInsert
            /// </summary>
            /// <param name="index"></param>
            /// <param name="value"></param>
            protected override void OnInsert(int index, LuogoDaVisitare value)
            {
                if (m_Percorso is object)
                    value.SetPercorso(m_Percorso);
                base.OnInsert(index, value);
            }

            /// <summary>
            /// OnSet
            /// </summary>
            /// <param name="index"></param>
            /// <param name="oldValue"></param>
            /// <param name="newValue"></param>
            protected override void OnSet(int index, LuogoDaVisitare oldValue, LuogoDaVisitare newValue)
            {
                if (m_Percorso is object)
                    newValue.SetPercorso(m_Percorso);
                base.OnSet(index, oldValue, newValue);
            }

            /// <summary>
            /// Carica dal db
            /// </summary>
            /// <param name="percorso"></param>
            protected internal virtual void Load(PercorsoDefinito percorso)
            {
                if (percorso is null)
                    throw new ArgumentNullException("percorso");
                Clear();
                
                m_Percorso = percorso;

                if (DBUtils.GetID(percorso, 0) == 0)
                    return;

                using (var cursor = new LuoghiDaVisitareCursor())
                {
                    cursor.Stato.Value = ObjectStatus.OBJECT_VALID;
                    cursor.IgnoreRights = true;
                    cursor.IDPercorso.Value = DBUtils.GetID(percorso, 0);
                    while (cursor.Read())
                    {
                        Add(cursor.Item);
                    }
                }
                     
                Sort();
                 
            }

            /// <summary>
            /// Imposta il percorso
            /// </summary>
            /// <param name="p"></param>
            protected internal void SetPercorso(PercorsoDefinito p)
            {
                m_Percorso = p;
                foreach (LuogoDaVisitare Item in this)
                    Item.SetPercorso(p);
            }
        }
    }
}

File: MiniDOMOffice/Office/Distribuzione/LuoghiDefinitiPerPercorso.vb
Imports FinSeA.Anagrafica
Imports FinSeA.Databases
Imports FinSeA.Sistema

Partial Class Office


    Public NotInheritable Class LuoghiDefinitiPerPercorso
        Inherits CCollection(Of LuogoDaVisitare)

        Private m_Percorso As PercorsoDefinito

        Public Sub New()
            Me.m_Percorso = Nothing
        End Sub

        Public Sub New(ByVal percorso As PercorsoDefinito)
            Me.New()
            Me.Load(percorso)
        End Sub

        Public ReadOnly Property Percorso As PercorsoDefinito
            Get
                Return Me.m_Percorso
            End Get
        End Property

        Protected Overrides Sub OnInsert(index As Integer, value As Object)
            If (Me.m_Percorso IsNot Nothing) Then DirectCast(value, LuogoDaVisitare).SetPercorso(Me.m_Percorso)
            MyBase.OnInsert(index, value)
        End Sub

        Protected Overrides Sub OnSet(index As Integer, oldValue As Object, newValue As Object)
            If (Me.m_Percorso IsNot Nothing) Then DirectCast(newValue, LuogoDaVisitare).SetPercorso(Me.m_Percorso)
            MyBase.OnSet(index, oldValue, newValue)
        End Sub

        Friend Sub Load(ByVal percorso As PercorsoDefinito)
            If (percorso Is Nothing) Then Throw New ArgumentNullException("percorso")
            Me.Clear()
            Me.m_Percorso = percorso
            If (GetID(percorso) <> 0) Then
                Dim cursor As New LuoghiDaVisitareCursor
                Try
                    cursor.Stato.Value = ObjectStatus.OBJECT_VALID
                    cursor.IgnoreRights = True
                    cursor.IDPercorso.Value = GetID(percorso)
                    While Not cursor.EOF
                        Me.Add(cursor.Item)
                        cursor.MoveNext()
                    End While
                Catch ex As Exception
                    Throw
                Finally
                    cursor.Dispose()
                End Try
                Me.Sort()
            End If
        End Sub

        Friend Sub SetPercorso(ByVal p As PercorsoDefinito)
            Me.m_Percorso = p
            For Each Item As LuogoDaVisitare In Me
                Item.SetPercorso(p)
            Next
        End Sub


    End Class
     
End Class